### PR TITLE
Fix AndroidJavaCompile.dataBindingDependencyArtifacts relocatability

### LINKED ISSUE #9: AndroidJavaCompile.dataBindingDependencyArtifacts is not relocatable
Android bug: https://issuetracker.google.com/issues/69243050

### PR DESCRIPTION
**Pull request by lptr**
Fixes #9.

### CODE REVIEW

File: src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy [hunk 1]
```diff
@@ -29,12 +29,13 @@ class AndroidCacheFixPlugin implements Plugin<Project> {
...
         new AndroidJavaCompile_AnnotationProcessorSource_Workaround(),
         new AndroidJavaCompile_ProcessorListFile_Workaround(),
+        new AndroidJavaCompile_DataBindingDependencyArtifacts_Workaround(),
         new ExtractAnnotations_Source_Workaround(),
         new IncrementalTask_CombinedInput_Workaround(),
         new StreamBasedTask_CombinedInput_Workaround(),
         new ProcessAndroidResources_MergeBlameLogFolder_Workaround(),
         new CheckManifest_Manifest_Workaround(),
-    ]
+    ] as List<Workaround>
```
Comment (lptr) [inline src/main/groovy/org/gradle/android/AndroidCacheFixPlugin.groovy]: IntelliJ complains if this is not here.

**Comment by lptr** [comment]
OK, I figured how to enable data binding. It seems to break everything, unfortunately :/

**Comment by lptr** [comment]
I'm not sure if this was worth it, but it seems it's fixed now.

**Comment by lptr** [comment]
Testing!

File: src/test/groovy/org/gradle/android/CompilerArgsProcessorTest.groovy [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package org.gradle.android
+
+import org.gradle.api.Project
+import spock.lang.Specification
+
+import static org.gradle.android.CompilerArgsProcessor.Skip
+import static org.gradle.android.CompilerArgsProcessor.SkipNext
+
+class CompilerArgsProcessorTest extends Specification {
```
Comment (wolfs) [inline src/test/groovy/org/gradle/android/CompilerArgsProcessorTest.groovy]: Could you add some tests for `AnnotationProcessorOverride`, too?

File: src/test/groovy/org/gradle/android/SimpleAndroidApp.groovy [hunk 3]
```diff
@@ -115,6 +115,11 @@ class SimpleAndroidApp {
             android {
                 compileSdkVersion 26
                 buildToolsVersion "26.0.2"
+                dataBinding.enabled = true
```
Comment (wolfs) [inline src/test/groovy/org/gradle/android/SimpleAndroidApp.groovy]: So the idea is that, even with databinding, most of the tasks are still loaded from cache, right?
Should we make that two tests, one with databinding enabled and one without?

File: src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            project.gradle.taskGraph.beforeTask {
+                if (task == it) {
+                    def processedArgs = processArgs(task, task.options.compilerArgs)
+                    overrideProperty(task, processedArgs)
+                }
+            }
+        }
+    }
+
+    List<String> processArgs(AndroidJavaCompile task, List<String> args) {
```
[the remark this answers is not in the document]
Comment (lptr) [reply src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy]: Retired.

File: src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                    def path = matcher.group(1)
+                    action.execute(task, path)
+                    break
+                }
+            }
+        }
+    }
+
+    static class Skip extends Rule {
+        Skip(String arg) {
```
[the remark this answers is not in the document]
Comment (lptr) [reply src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy]: Obliterated.

File: src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            return new Skip(Pattern.compile(pattern))
+        }
+
+        @Override
+        void process(Matcher match, Collection<String> processedArgs, Iterator<String> remainingArgs) {
+        }
+    }
+
+    static class SkipNext extends Rule {
+        SkipNext(String arg) {
```
[the remark this answers is not in the document]
Comment (lptr) [reply src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy]: Death to it!

File: src/test/groovy/org/gradle/android/SimpleAndroidApp.groovy (hunk 3 above)
Comment (lptr) [reply src/test/groovy/org/gradle/android/SimpleAndroidApp.groovy]: Good idea. Let's do that when we come back to this PR (if at all necessary).

File: src/test/groovy/org/gradle/android/CompilerArgsProcessorTest.groovy (hunk 2 above)
Comment (lptr) [reply src/test/groovy/org/gradle/android/CompilerArgsProcessorTest.groovy]: OK.

File: src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            if (remainingArgs.hasNext()) {
+                remainingArgs.next()
+            }
+        }
+    }
+
+    static abstract class Rule {
+        final Pattern pattern
+
+        Rule(String arg) {
```
Comment (wolfs) [inline src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy]: Is this constructor now unused?

Comment (lptr) [reply src/main/groovy/org/gradle/android/CompilerArgsProcessor.groovy]: Take a look now please

**Comment by lptr** [comment]
Let's put this in the icebox until we can confirm that we actually need this.

**Pull request merged**